compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: compartmentalize-grains-via pop
+- Start Date: 2020-06-22
+- SEP Status: Draft
+- SEP PR: (leave this empty)
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Grains is currently a static component of salt.  We want it to be developed in a more compartmentalized fashion.
+The engineers at saltstack have created kernel-specific projects that functionally replace the grains
+provided by salt with POP.  We propose that this project, codenamed `corn` replaces grains in salt while
+maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+"The original implementation of grains was not designed to scale to the level that salt requires" - thatch45
+Tom wrote POP to solve this problem and we're going to use it. 
+
+By using Plugin Oriented Programming, grains can be separated into individual projects that can be maintained
+by the community safely.
+
+This also solves the issue of `core.py` being an unmaintainable monolith.  It allows platform specific grains
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual grains projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to and tested on the new platform.
+- The new platform has many more features than the salt-grains implementation, but is compatible with salt grains.
+- None of the grain names have been changed, but the output of grains is more consistent/complete between platforms.
+- This will resolve many bugs that are open against grains and will migrate those bugs OUT of the salt
+platform -- making salt easier to develop
+- Grains in the new platform are independently and fully tested.
+- Most grains tests will be able to be removed from salt; speeding up it's test suite.
+- This comprises a complete rewrite of ALL grains.  Backwards compatibility is expected, but we need this to get
+in early so that any unintentional issues can be resolved quickly.
+- Salt releases will be able to pin idem-grain dependencies with each release.
```

Comment:
What does "idem" mean in this context?

Reply:
It also opens the door to future exec modules and states to be added to idem-windows and other platform-specific projects.  That way they you don't need to worry about tests for all OSes passing when your exec module only affects windows.

Reply:
That doesn't answer the question. Why is the Latin for "same" being used? I assume it doesn't mean "same-grain" or it would just say that.

Reply:
`[idem](https://gitlab.com/saltstack/pop/idem)` is short for "idempotent".  It is the evolution of the execution and state module system from salt

Reply:
This isn't a poppadom/pop-idem pun is it? \*squints\*

Reply:
idempotent
[ ahy-duh m-poht-nt, id-uh m- ] 
(computing) Describing an action which, when performed multiple times, has no further effect on its subject after the first time it is performed.